model/riscv_insts_vext_arith.sail

```diff
@@ -0,0 +1,554 @@
+  VV_VSADDU        <-> 0b100000,
+  VV_VSADD         <-> 0b100001,
+  VV_VSSUBU        <-> 0b100010,
+  VV_VSSUB         <-> 0b100011,  
+  VV_VSLL          <-> 0b100101,
+  VV_VSMUL         <-> 0b100111,
+  VV_VSRL          <-> 0b101000,
+  VV_VSRA          <-> 0b101001,
+  VV_VSSRL         <-> 0b101010,
+  VV_VSSRA         <-> 0b101011
+}
+
+mapping clause encdec = VVTYPE(funct6, vm, vs2, vs1, vd)
+  <-> encdec_vvfunct6(funct6) @ vm @ vs2 @ vs1 @ 0b000 @ vd @ 0b1010111
+
+function clause execute(VVTYPE(funct6, vm, vs2, vs1, vd)) = {
+  let SEW_pow  : int = get_sew_pow();
+  let SEW      : int = int_power(2, SEW_pow);
+  let LMUL_pow : int = get_lmul_pow();
+  let VLEN_pow : int = get_vlen_pow();
+  let num_elem : int = get_num_elem(LMUL_pow, SEW);
+
+  assert(SEW == 8 | SEW == 16 | SEW == 32 | SEW == 64);
+  let 'n = num_elem;
+  let 'm = SEW;
+
+  let vm_val  : vector('n, dec, bool)     = read_vmask(num_elem, vm, vreg_name("v0"));
+  let vs1_val : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, LMUL_pow, vs1);
+  let vs2_val : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, LMUL_pow, vs2);
+  let vd_val  : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, LMUL_pow, vd);
+  result      : vector('n, dec, bits('m)) = undefined;
+  mask_helper : vector('n, dec, bool)     = undefined;
+
+  (result, mask_helper) = init_masked_result(num_elem, SEW, LMUL_pow, vd_val, vm_val);
+
+  foreach (i from 0 to (num_elem - 1)) {
+    if mask_helper[i] == true then {
+      result[i] = match funct6 {
+        VV_VADD          => vs2_val[i] + vs1_val[i],
+        VV_VSUB          => vs2_val[i] - vs1_val[i],
+        VV_VAND          => vs2_val[i] & vs1_val[i],
+        VV_VOR           => vs2_val[i] | vs1_val[i],
+        VV_VXOR          => vs2_val[i] ^ vs1_val[i],
+        VV_VSADDU        => unsigned_saturation('m, EXTZ('m + 1, vs2_val[i]) + EXTZ('m + 1, vs1_val[i])),
+        VV_VSADD         => signed_saturation('m, EXTS('m + 1, vs2_val[i]) + EXTS('m + 1, vs1_val[i])),
+        VV_VSSUBU        => {
+                              if unsigned(vs2_val[i]) < unsigned(vs1_val[i]) then zeros()
+                              else unsigned_saturation('m, EXTZ('m + 1, vs2_val[i]) - EXTZ('m + 1, vs1_val[i]))
+                            },
+        VV_VSSUB         => signed_saturation('m, EXTS('m + 1, vs2_val[i]) - EXTS('m + 1, vs1_val[i])),
+        VV_VSMUL         => {
+                              result_mul = to_bits('m * 2, signed(vs2_val[i]) * signed(vs1_val[i]));
+                              rounding_incr = get_fixed_rounding_incr(result_mul, 'm - 1);
+                              result_wide = (result_mul >> ('m - 1)) + EXTZ('m * 2, rounding_incr);
+                              signed_saturation('m, result_wide['m..0])
+                            },
+        VV_VSLL          => {
+                              let shift_amount = get_shift_amount(vs1_val[i], SEW);
+                              assert(shift_amount >= 0);
```

Comment:
I tried to delete this assert statement and then the next line `vs2_val[i] << shift_amount` will cause an error:

> 84 |                              vs2_val[i] << shift_amount
   |                              ^------------------------^
   | No overloading for (operator <<), tried:
   | * shift_bits_left
   |    Could not unify bitvector('fv59161#m, dec) and int('_shift_amount)
   | * shiftl
   |    Could not resolve quantifiers for shiftl
   |    * '_shift_amount >= 0

so this assert statement is necessary for the compilation.

Reply:
If you change the type of `get_shift_amount` to return `nat` the assert will not be necessary.

Reply:
Yep, I think the return type of get_shift_amount should either nat. If you need a more complex return type, you can also use something like `{'n, 0 <= 'n <= upper_bound. int('n)}` or similar.

Functions are type-checked individually, so if you say get_shift_amount returns an integer, then it will assume it can be negative in functions using get_shift_amount.

Reply:
Got it. I'll modify that return type. Thanks!

Reply:
This assert statement has been removed.